Autodetect TestFramework if none is configured explicitly

marcphilipp opened the pull request:
When the TestFramework of a Test task is not configured explicitly it
used to default to `useJUnit()`. Now, the task's classpath is inspected
for a junit-platform-engine.jar or testng.jar. If one of them is found,
the task uses the JUnit Platform or TestNG to execute tests,
respectively.

Since we have internal dependencies on the JUnit Platform and TestNG,
all Test tasks in our own build are configured to `useJUnit()` and
explicitly.

marcphilipp commented:
@big-guy Where should the notes for this change go? `upgrading_version_5.adoc` or `notes.md`?

oehme commented on subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/TestFrameworkAutoDetection.java:
```diff
@@ -0,0 +1,67 @@
+    private static Class<? extends TestFramework> detectTestFrameworkFromClasspath(Test task) {
+        final MutableReference<Class<? extends TestFramework>> testFramework = MutableReference.empty();
+        testFramework.set(JUnitTestFramework.class);
+        task.getClasspath().getAsFileTree().visit(new EmptyFileVisitor() {
```
⭕️  Making it a file tree seems wasteful, as we don't care about children of directories. Why not visit the file collection directly?

marcphilipp replied:
Good catch! 👍

marcphilipp commented:
I wasn't aware groovy-all now pulls in JUnit Jupiter _by default_:
https://scans.gradle.com/s/56m5oaqyt7gpq/dependencies?focusedDependency=WzAsNiw3NCxbMCw2LFswLDY1LDc0XV1d&toggled=W1swXSxbMCw2XSxbMCw2LFswXV0sWzAsNixbMCw2NV1dLFswLDYsWzAsNjUsNzRdXV0

Now, I'm worried about breaking a lot of Groovy projects which use JUnit 4 and/or Spock and would now have to notice that none of their tests are still executed and be forced to add `useJUnit()`. I tend to withdraw this PR as it might do more harm than good. @britter Any ideas?

oehme commented:
You could give junit 4 precedence unless the vintage engine is also present. That way you get auto detection but more in line with the previos behavior.

britter commented on subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/TestFrameworkAutoDetection.java:
```diff
@@ -0,0 +1,62 @@
+                        testFramework = JUnitPlatformTestFramework.class;
+                        break;
+                    } else if ("testng".equals(artifactName)) {
+                        testFramework = TestNGTestFramework.class;
```
Is the `break` statement intentionally missing here?

marcphilipp replied:
Yes, to give junit-platform-engine precedence over testng, should it be found in a later classpath entry.

lacasseio commented:
> @big-guy Where should the notes for this change go? `upgrading_version_5.adoc` or `notes.md`?

It seems to be a potential breaking change so it would go in `upgrading_version_5.adoc`. If the impact of selecting a test framework is a nice thing to have, then maybe another note in `notes.md` to announce the "feature" would also be a good idea.

marcphilipp commented:
> You could give junit 4 precedence unless the vintage engine is also present. That way you get auto detection but more in line with the previous behavior.

That works. I added a few additional commits and did the same for TestNG (which also comes with groovy-all). Before I fix the few remaining test failures, let's decide whether we want this change or not.

> We're tying ourselves to particular _file names_ to detect which testing framework is used vs something expressed in the metadata of the dependencies or particular classes (that we know must exist).

Right, I looked into the latter option first but it would require a lot of work to expose that to the `Test` task, in particular, if users set `classpath` manually. However, as a heuristic I think relying on the file names is fine. Users can always configure it explicitly.

> We're introducing a subtle resolve-at-configuration-time behavior where sometimes the configuration for test dependencies will be resolved if someone calls `getTestFramework` too early.

I don't see a reason why anyone would call `getTestFramework()`. If they did, they'd get a deprecation warning for resolving a configuration at configuration time, wouldn't they?

> We're keeping useful information (do we have JUnit tests, TestNG tests, JUnit5 tests, etc) in a task's configuration.

I don't quite understand this concern. We already do this today, don't we? Could you please elaborate?

> Other than convenience, what else would a user get out of this?  

It's about providing a sensible default. More and more new projects start out using the JUnit Platform instead of JUnit 4 and I'd like to remove the burden of having to tinker with the `Test` task's configuration for them.

big-guy commented on subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java:
```diff
@@ -820,10 +824,15 @@ public Test setExcludes(Iterable<String> excludes) {
 
+    @Deprecated
     public TestFramework testFramework(@Nullable Closure testFrameworkConfigure) {
+        DeprecationLogger.nagUserOfDiscontinuedMethod("testFramework", "Please call useJUnit(), useJUnitPlatform(), or useTestNG() instead.");
```
👍

big-guy commented on subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java:
```diff
@@ -820,10 +824,15 @@ public Test setExcludes(Iterable<String> excludes) {
 
     @Internal
     public TestFramework getTestFramework() {
```
I talked to Adam about my misgivings on this and I think there are a couple of things...

- The default/convention for determining the test framework to use is in the Test task itself vs pushing out to the plugins.
- The way this is wired into up-to-date checks looks brittle.  This method just happens to get called in the right place (`getOptions()` below) to cause the testFramework to be set. 
- The way we detect the test framework is brittle as well (based on file names).  
- I think this also pushes us further away from something higher-level from deciding which test framework to use.  e.g., you could imagine a plugin that would add the JUnit dependency and could set the test framework too.

I've pushed this commit to a separate branch: 
https://github.com/gradle/gradle/commit/86a1e971699d20d4855c626e9746cf508fc2716e
- On Adam's suggestion, I changed testFramework into a `Property<TestFramework>` and exposed it as another getter. I didn't do it, but this getter could be marked as `@Nested` and the TestFramework subclasses could be annotated properly to make it all work. This would make it clearer that the TestFramework was an input of some kind and prevent anything from sneaking in as an input. I don't think we need to do this now.
- I pulled out the details for creating TestFramework instances into a factory and moved the way the conventional TestFramework is decided.  In each of our plugins that create Test tasks, I applied some convention.  In JavaBasePlugin, the convention is to use JUnit (what we do now).  In JavaPlugin, the convention is to use the test framework detection to pick a test framework.  In the JUnit software model plugin, the convention is to use JUnit always.
- To avoid sniffing file names, I also changed the test framework detection to look at the _direct_ dependencies of the test's implementation.  I don't think this is as good as it could be (since I see multiple JUnit platform dependencies that look roughy equivalent), but it follows a simple rule: if JUnit platform is available, use JUnit platform.  If TestNG is available, use TestNG. if JUnit is available, use JUnit.  If nothing is available, assume JUnit (for backwards compatibility). 

I think this alleviates my initial concerns:
- We can't resolve dependencies accidentally during configuration
- We avoid inspecting file names 
- We don't make the Test task more complicated (it just does whatever testFramework says).

My commit breaks some tests, but it seems to work mostly.

big-guy commented on subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/TestFrameworkAutoDetection.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class TestFrameworkAutoDetection {
```
I have a longer comment below, but I took a stab at rewriting this to be based on the _direct_ dependencies of the tests:

https://github.com/gradle/gradle/commit/86a1e971699d20d4855c626e9746cf508fc2716e#diff-44b9afd6ec27aaa2a010c1ac97c75191

I think there should be a simple rule (e.g., we prefer this over that), even if sometimes we don't suss out exactly what the user intended.   I think I don't quite have it right because it looks like junit-vintage-engine should select junit platform sometimes.

Maybe we'd also want to make spockframework auto-detect as JUnit.

marcphilipp replied:
> I think there should be a simple rule (e.g., we prefer this over that), even if sometimes we don't suss out exactly what the user intended. I think I don't quite have it right because it looks like junit-vintage-engine should select junit platform sometimes.

I tried a simple rule initially but it turned to be more complicated (see discussion above wrt. groovy-all). You shouldn't check for `org.junit.jupiter:junit-jupiter` because Jupiter != Platform and that wouldn't work if someone used a different engine (there are actually quite a few by now). Instead, the check should be for `org.junit.platform:junit-platform-engine` which is the base dependency for any engine but use JUnit 4 if `junit:junit` is present unless `org.junit.vintage:junit-vintage-engine` is also present. For TestNG it's similar because groovy-all also includes a dependency on it and we still should default to JUnit 4 if it's also present.

Thus, I don't think it's going to get much simpler than this (using coordinates instead of file names, of course):

```java
if (artifacts.contains("junit-platform-engine")
        && (!artifacts.contains("junit") || artifacts.contains("junit-vintage-engine"))) {
    // Use JUnit Platform
} else if (artifacts.contains("testng") && !artifacts.contains("junit")) {
    // Use TestNG
} else {
    // Use JUnit
}
```

> Maybe we'd also want to make spockframework auto-detect as JUnit.

I wouldn't do that because Spock 2.0 will use the JUnit Platform.

big-guy commented on subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestFrameworkAutoDetectionIntegrationTest.groovy:
```diff
@@ -0,0 +1,136 @@
+import org.gradle.testing.fixture.JUnitCoverage
+import org.gradle.testing.fixture.TestNGCoverage
+
+class TestFrameworkAutoDetectionIntegrationTest extends AbstractIntegrationSpec {
```
The dependencies used in these tests are sort of confusing. 

In some cases, we're adding `junit-jupiter`, which appears to be the right dependency for JUnit 5+, but sometimes we're adding junit-jupiter-engine and we're almost always adding testng.  And we're mostly adding dependencies to testImplementation, but our samples use a wildly different mix.

I think these tests should assert the test framework used vs relying on just the tests passing/failing.  I think that could be done by adding a `doLast` to the test task and asserting testFramework has a particular type.

marcphilipp replied:
IIRC `testng` is always there to show that both JUnit Platform and JUnit take precedence.

big-guy commented on subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestFrameworkAutoDetectionIntegrationTest.groovy:
```diff
@@ -0,0 +1,136 @@
+        assertTestPassed('JupiterTestClass')
+    }
+
+    def "uses JUnit Platform if JUnit 4 and junit-vintage-engine are on the classpath"() {
```
I was just poking around with auto test detection and was able to make this still test pass when using the wrong test framework (JUnit vs JUnitPlatform).

marcphilipp replied:
You're right. This should also include a test for another engine or simply check for the `TestFramework` being used.

big-guy commented on subprojects/core/src/integTest/groovy/org/gradle/api/ApplyPluginIntegSpec.groovy:
```diff
@@ -199,6 +199,10 @@ class ApplyPluginIntegSpec extends AbstractIntegrationSpec {
+
+            test {
+                useJUnit()
+            }
```
Shouldn't the default still be JUnit?

marcphilipp replied:
Yes, I think was an intermediate change I did while trying to fix tests. I think it can be reverted.

big-guy commented on subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/JavaProjectInitDescriptor.java:
```diff
@@ -105,9 +102,6 @@ private void addTestFramework(BuildInitTestFramework testFramework, BuildScriptB
                     ).testRuntimeOnlyDependency(
                     "Use JUnit Jupiter Engine for testing.",
                     "org.junit.jupiter:junit-jupiter-engine:" + libraryVersionProvider.getVersion("junit-jupiter")
```
If junit-jupiter is new enough, could we switch to the simpler declaration?

marcphilipp replied:
Yes, `testImplementation("org.junit.jupiter:junit-jupiter")` is enough but it did not exist prior to `5.4.0`.

big-guy commented on subprojects/docs/src/samples/testing/junitplatform/mix/groovy/build.gradle:
```diff
@@ -29,7 +29,3 @@ dependencies {
     testRuntimeOnly 'org.junit.vintage:junit-vintage-engine:5.1.0'
```
Can you sanity check the dependencies in our samples?  It looks like we could upgrade to a newer version of junit-jupiter and make the sample simpler.

big-guy commented on subprojects/docs/src/docs/userguide/java_testing.adoc:
```diff
@@ -289,9 +289,9 @@ Unlike its predecessor, JUnit 5 is modularized and composed of several modules:
  and http://junit.org/junit5/docs/current/user-guide/#extensions[extension model] for writing tests and extensions in JUnit 5. JUnit Vintage provides a `TestEngine` for running JUnit 3 and JUnit 4 based tests on the platform.
 
-The following code enables JUnit Platform support in `build.gradle`:
+Gradle detects whether the JUnit Platform is on the classpath of the `Test` task and automatically enables its built-in JUnit Platform support. If you don't want to rely on the autodetection or configure it explicitly, you can enable it in your buildscript:
```
```suggestion
Gradle can automatically detect if the JUnit Platform can be used with a `Test` task. If you don't want to rely on the auto-detection or need to explicitly configure it, you can enable JUnit Platform support in your buildscript:
```

marcphilipp commented:
> On Adam's suggestion, I changed testFramework into a `Property<TestFramework>` and exposed it as another getter. I didn't do it, but this getter could be marked as `@Nested` and the TestFramework subclasses could be annotated properly to make it all work. This would make it clearer that the TestFramework was an input of some kind and prevent anything from sneaking in as an input. I don't think we need to do this now.

I like it and agree that we don't need to make it an input for now.

> I pulled out the details for creating TestFramework instances into a factory and moved the way the conventional TestFramework is decided.  In each of our plugins that create Test tasks, I applied some convention.  In JavaBasePlugin, the convention is to use JUnit (what we do now).  In JavaPlugin, the convention is to use the test framework detection to pick a test framework.  In the JUnit software model plugin, the convention is to use JUnit always.

👍 

> To avoid sniffing file names, I also changed the test framework detection to look at the _direct_ dependencies of the test's implementation.  I don't think this is as good as it could be (since I see multiple JUnit platform dependencies that look roughy equivalent), but it follows a simple rule: if JUnit platform is available, use JUnit platform.  If TestNG is available, use TestNG. if JUnit is available, use JUnit.  If nothing is available, assume JUnit (for backwards compatibility). 

I think that's good. I originally didn't do it this way because users might set the `classpath` on the `Test` task explicitly and it would no longer be connected to `testImplementation`. But I think that should be fine since one can always call `useJUnit()` etc. explicitly.

> I've pushed this commit to a separate branch: 
https://github.com/gradle/gradle/commit/86a1e971699d20d4855c626e9746cf508fc2716e

Feel free to merge it with this branch or open a new PR that supersedes this one.

> Can you also add something to the release notes.md and release-features.txt?

Please let me know if i should add it to this branch or yours.

marcphilipp commented:
@big-guy and I discussed this PR offline. In my opinion only checking direct dependencies isn't enough because we would not be able to support arbitrary test engines. On the other hand, his counterargument against taking into account transitive dependencies is a good one: adding a dependency might change the selected test framework and go unnoticed. To avoid this, we could make the task fail if the test framework is not set explicitly and cannot be auto-detected unambiguously. However, that would cause a lot of projects to fail when upgrading to Gradle 6.0 and does not seem worth it. Therefore, I'm closing this PR.
